<a href="https://colab.research.google.com/github/Albiorix2016/Spring-2023-Discrete-Mathematics-CPSMA-3913-01-/blob/main/Module6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

We were asked to perform the following:
1. Compute using modulo arithmetic

(a) Convert the number 240367 into base ten.

(b) Convert the hexadecimal number 72e3 or 0x72e3 into base ten.

(c) Convert 3735928559 into hexadecimal.

(d) Using two’s complement preform the base ten subtraction 25 − 37

And then check our work. 

I wrote the following in python to help with this.

First: Convert 240367 into base ten

In [7]:
# (a) Convert the number 240367 into base ten.
print(int(240367))

240367


Convert the number 240367 into base ten. The answer is 386344.

Second: Convert the hexadecimal number 72e3 or 0x72e3 into base ten

In [8]:
# (b) Convert the hexadecimal number 72e3 or 0x72e3 into base ten.
print("0x72e3 in base ten is:", 7*16**3 + 2*16**2 + 14*16 + 3)

0x72e3 in base ten is: 29411


Convert the hexadecimal number 72e3 or 0x72e3 into base ten. The answer is 29411.

Third: Convert 3735928559 into hexadecimal

In [9]:
# (c) Convert 3735928559 into hexadecimal.
print("3735928559 in hexadecimal is:", hex(3735928559))

3735928559 in hexadecimal is: 0xdeadbeef


Final asked to use two's complement perform the base then subtraction 25-37 

In [ ]:
# (d) Using two’s complement perform the base ten subtraction 25 − 37. Check your answer.
# Convert 25 and 37 to binary (using 6 bits for this example)
a = bin(25 & 0b111111)  # 0b11001
b = bin(37 & 0b111111)  # 0b100101
# Perform two's complement subtraction
diff = (int(a, 2) + int(b, 2) ^ 0b111111 + 1) & 0b111111
# Convert the result back to decimal
result = diff - 64 if diff > 31 else diff
# Check the answer
print("25 - 37 = ", result)